### PR TITLE
Retrieve the query identifier in profiler output.

### PR DESCRIPTION
**Pull request by rjuju**
As mentioned in https://github.com/okbob/plpgsql_check/issues/79 please find an initial attempt to handle queryid in the profiler.

Note that I had to add a few macro due to the queryid change in pg11.  I also didn't use a default case in the profiler_get_queryid switch, to make sure we can get a warning if a new statement type is added.  I tried to cover all possible statement types, but as I'm not very familiar with plpgsql source code I may have  missed something.

If multiple plans are found, only the first queryid found is used, I don't know if that's ok.

I arbitrarily put the queryid as the first column, I can move it somewhere else if you prefer.

The queryid is cached in the profiler_stmt to avoid retrieving it more than necessary.  I can add a new GUC to enable/disable queryid processing if needed.

### CODE REVIEW

File: plpgsql_check--1.13.sql [hunk 1]
```diff
@@ -139,7 +139,8 @@ AS $$
 $$ LANGUAGE sql;
 
 CREATE FUNCTION plpgsql_profiler_function_tb(funcoid regprocedure)
-RETURNS TABLE(lineno int,
+RETURNS TABLE(queryids int8[],
+              lineno int,
```
Comment (okbob) [inline plpgsql_check--1.13.sql:143]: There is probably bug - should be only int8 not int8[].

File: src/format.c [hunk 2]
```diff
@@ -71,35 +71,37 @@ static void put_error_tabular(plpgsql_check_result_info *ri, PLpgSQL_execstate *
...
-#define Anum_profiler_cmds_on_row		2
-#define Anum_profiler_exec_count		3
-#define Anum_profiler_total_time		4
-#define Anum_profiler_avg_time			5
-#define Anum_profiler_max_time			6
-#define Anum_profiler_processed_rows	7
-#define Anum_profiler_source			8
+#define Natts_profiler					10
+
+#define Anum_profiler_queryid			0
```
Comment (okbob) [inline src/format.c:76]: I don't think so queryid should be first column - probably next after Anum_profiler_stmt_lineno

File: src/profiler.c [hunk 3]
```diff
@@ -1477,6 +1482,128 @@ profiler_touch_stmts(profiler_info *pinfo,
 	}
 }
 
+/* Return the first queryid found in the given PLpgSQL_stmt, if any. */
+static pc_queryid
+profiler_get_queryid(PLpgSQL_stmt *stmt)
```
Comment (okbob) [inline src/profiler.c:1487]: please, divide this routine to two - one returns expression assigned to statement, second returns queryid from expression

File: src/profiler.c [hunk 4]
```diff
@@ -1477,6 +1482,128 @@ profiler_touch_stmts(profiler_info *pinfo,
...
+			break;
+		case PLPGSQL_STMT_RETURN_QUERY:
+			{
+				PLpgSQL_stmt_return_query *q;
+
+				q = (PLpgSQL_stmt_return_query *) stmt;
+				if (q->query)
+					expr = q->query;
+				else
+					expr = q->dynquery;
```
Comment (okbob) [inline src/profiler.c:1545]: There is a questions how to solve well dynamic queries. In this case dynquery probably will be pretty fast, but for you is interesting evaluated dynquery

File: plpgsql_check--1.13.sql (hunk 1 above)
Comment (okbob) [reply plpgsql_check--1.13.sql:143]: I see it now - using the array is correct.

Comment (rjuju) [reply plpgsql_check--1.13.sql:143]: There can be multiple instruction per line, so we need to track all of them.  For instance:

```
# select queryids,lineno, source from plpgsql_profiler_function_tb('f1');
                  queryids                   | lineno |        source         
---------------------------------------------+--------+-----------------------
 <NULL>                                      |      1 | 
 <NULL>                                      |      2 | declare
 <NULL>                                      |      3 |     i integer;
 {NULL}                                      |      4 | begin
 {-2184263057854310049,-5581434024648880688} |      5 |     i := 1; i := i+1;
 <NULL>                                      |      6 | end;
(6 rows)

```

File: src/format.c (hunk 2 above)
Comment (rjuju) [reply src/format.c:76]: Sure, I'll fix it!

File: src/profiler.c (hunk 3 above)
Comment (rjuju) [reply src/profiler.c:1487]: Agreed.

File: src/plpgsql_check.h [hunk 5]
```diff
@@ -7,6 +7,17 @@
 #include "access/tupdesc.h"
 #include "storage/ipc.h"
 
+#if PG_VERSION_NUM >= 110000
+typedef uint64 pc_queryid;
+#define NOQUERYID				(UINT64CONST(0))
+#define QueryidGetDatum(datum)	(UInt64GetDatum(datum))
```
Comment (okbob) [inline src/plpgsql_check.h:13]: I don't think so introduction QueryidGetDatum aliases for UInt is good idea. Queryid is not custom type, and using naming convention related to custom types is messy. Just use int64 everywhere - and on pre and 11 use cast to int64. I use typdef only for composites, so this typdef is not consistent with other code

File: src/profiler.c (hunk 4 above)
Comment (rjuju) [reply src/profiler.c:1545]: I don't understand your point here.  According to `exec_stmt_return_query()` only one of those can be valued.

File: plpgsql_check--1.13.sql (hunk 1 above)
Comment (okbob) [reply plpgsql_check--1.13.sql:143]: yes, this is true

File: src/plpgsql_check.h (hunk 5 above)
Comment (rjuju) [reply src/plpgsql_check.h:13]: I tried that approach first, but there's no UInt64GetDatum()  in pg 9.5, so I went with that instead.  I guess I could use Int64GetDatum() instead in every case?

File: src/profiler.c (hunk 4 above)
Comment (okbob) [reply src/profiler.c:1545]: yes, but dynquery is (if I remember well) is expression like 'SELECT * FROM ' || variable. Probably queryid of this expression is useless - interesting is queryid of evaluated dynquery - this query that will be executed, and that can be slow

Comment (okbob) [reply src/profiler.c:1545]: <pre>
<--><-->Assert(stmt->dynquery != NULL);
<--><-->query = exec_eval_expr(estate, stmt->dynquery,
<--><--><--><--><--><--><-->   &isnull, &restype, &restypmod);
<--><-->if (isnull)
<--><--><-->ereport(ERROR,
<--><--><--><--><-->(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
<--><--><--><--><--> errmsg("query string argument of EXECUTE is null")));

<--><-->/* Get the C-String representation */
<--><-->querystr = convert_value_to_string(estate, query, restype);

<--><-->/* copy it into the stmt_mcontext before we clean up */
<--><-->querystr = MemoryContextStrdup(stmt_mcontext, querystr);

<--><-->exec_eval_cleanup(estate);

<--><-->/* Execute query, passing params if necessary */
<--><-->rc = SPI_execute_with_receiver(querystr,
<--><--><--><--><--><--><--><--><-->   exec_eval_using_params(estate,
<--><--><--><--><--><--><--><--><--><--><--><--><--><--><-->  stmt->params),
<--><--><--><--><--><--><--><--><-->   estate->readonly_func,
<--><--><--><--><--><--><--><--><-->   0,
<--><--><--><--><--><--><--><--><-->   treceiver);
</pre>

File: src/plpgsql_check.h (hunk 5 above)
Comment (okbob) [reply src/plpgsql_check.h:13]: If you use one or second, there will be some inconsistency - I usually prefer maximum similarity with related Postgres code. 9.5 is not too important - it is near to EOL, but until it is supported, plpgsql_check should be compiled without warnings.

Comment (rjuju) [reply src/plpgsql_check.h:13]: I'm a bit confused here.   Should I use Int64GetDatum() in all code path, with an explicit cast for version 10 and before?

File: src/profiler.c (hunk 4 above)
Comment (rjuju) [reply src/profiler.c:1545]: Ah I see, the code as is will return the queryid of the query that generates the query string to be executed, which is totally useless.

Indeed, if we want to get the real queryid, we'd need to do also generate the query string and then call up to parse analysis to get the queryid.  That seems like a lot of overhead, and also a lot of complication as exec_eval_expr isn't exported.

Comment (okbob) [reply src/profiler.c:1545]: There is one situation, where you use dynquery - when dynquery is constant. It is common pattern for forcing one shot plan in plpgsql (workaround when generic plan is suboptimal). Elsewhere I afraid so queryid is impossible - more, for dynamic SQL the queryid is not stable. You can see `check_dynamic_sql`. Simillar trick you can use for check if expression is constant, and then you can calculate queryid from this string. Still you should to calculate queryid by self.

Comment (rjuju) [reply src/profiler.c:1545]: Indeed, it should be possible to work around limitation for constant expression.  But if the expression is dynamic, I don't see any simple way to compute the queryid, as I don't see how to generate the string?

Comment (okbob) [reply src/profiler.c:1545]: I don't think so there is solution on plpgsql_check level. You can write patch to postgres and introduce SPI to be possible returns last queryid - like now is returned state, or row count. This can clean current patch (less access to lower structures), and can help with dynamic queries too. For this moment there are two strategies: 1) check if expression is immutable, and if it, you can evaluate expression second time inside plpgsql_check, 2) do nothing - just returns -1 or some special number. But with first strategy there is a possibility to get lot of different queryids - so there should be some algorithm like reduce it - maybe queryid of the most slow query.

File: src/plpgsql_check.h (hunk 5 above)
Comment (okbob) [reply src/plpgsql_check.h:13]: yes

File: src/profiler.c (hunk 4 above)
Comment (okbob) [reply src/profiler.c:1545]: There was some patch for introduction queryid to SPI and nested queries to SPI. It can be used for inspiration.

There is high probability so expression of dynamic SQL will be simple - and then you can evaluate second time. The overhead will be minimal. It is workaround - better is system solution on SPI level. If you send patch to commitfest, I'll do review and support.

Comment (rjuju) [reply src/profiler.c:1545]: Well, sure but I was first targeting support for existing release.  For core changes, I was thinking copying the queryid around for dynamic queries, but exposing it in the state is a way better id!

Few points:

> you can evaluate expression second time inside plpgsql_check

IIUC the only way to do that would be to construct a whole function containing the expression and evaluate it, as none of the infrastructure below `plpgsql_exec_function()` is exposed, or did I miss something?

> there is a possibility to get lot of different queryids

Do you mean that a single expression could yield multiple queryid?

> There was some patch for introduction queryid to SPI and nested queries to SPI. It can be used for inspiration.

Do you have any pointer on that discussion?

Comment (okbob) [reply src/profiler.c:1545]: > Well, sure but I was first targeting support for existing release. For core changes, I was thinking copying the queryid around for dynamic queries, but exposing it in the state is a way better id!

sure

> 
> Few points:
> 
> > you can evaluate expression second time inside plpgsql_check
> 
> IIUC the only way to do that would be to construct a whole function containing the expression and evaluate it, as none of the infrastructure below `plpgsql_exec_function()` is exposed, or did I miss something?

My original idea was wrong - the necessary API is static (is not visible). But fortunately there is a plugin API, that is used by profiler too, and this API provides function *assign_expr - you can look to `plpgsql_check_trace_assert_on_stmt_beg`. There is used for evaluating `ASSERT` expressions.

> 
> > there is a possibility to get lot of different queryids
> 
> Do you mean that a single expression could yield multiple queryid?

sure - developers are idiots also - I seen the code `EXECUTE 'SELECT * FROM users WHERE name = ' || quote_ident(var)`. I think you can remember just only one and time, and display for result just one of most slower query.

> 
> > There was some patch for introduction queryid to SPI and nested queries to SPI. It can be used for inspiration.
> 
> Do you have any pointer on that discussion?

https://www.postgresql.org/message-id/CABUevEzzwefAfJbJKvwYfJ73gp9SJd-5WAF-oNiVuY0%2BUjW%3DQw%40mail.gmail.com

https://www.postgresql.org/message-id/CAOBaU_YsiPHJ-mm26omMY7hgBBXfx9Ld0xw-DWrdvpbBKpbqQg@mail.gmail.com

Probably we need little bit different access - on SPI level - the queries are nested and you need to get queryid on correct level.

Comment (rjuju) [reply src/profiler.c:1545]: > But fortunately there is a plugin API, that is used by profiler too, and this API provides function *assign_expr - you can look to plpgsql_check_trace_assert_on_stmt_beg. There is used for evaluating ASSERT expressions.

Oh great, that indeed should be enough.  I'll do some testing to see if there are caveats.  If that works there should be a new GUC to avoid wasting time computing queryid if users aren't interested in that.

>  I seen the code EXECUTE 'SELECT * FROM users WHERE name = ' || quote_ident(var). I think you can remember just only one and time, and display for result just one of most slower query.

Ah I see, you mean that there'd be a qeryid for the generated SELECT statement *and* a queryid for the quote_ident() call?

I didn't look thoroughly at plpgsql implementation yet, but I'm assuming that if we add an infrastructure to report some SPI call's queryid, it should only return the top level one, so the nested one will be automatically discarded by callers like plpgslq_check.

Comment (okbob) [reply src/profiler.c:1545]: > > I seen the code EXECUTE 'SELECT * FROM users WHERE name = ' || quote_ident(var). I think you can remember just only one and time, and display for result just one of most slower query.
> 
> Ah I see, you mean that there'd be a qeryid for the generated SELECT statement _and_ a queryid for the quote_ident() call?

I don't understand - theoretically you need queryid for any possible slow query. Building expression (building SELECT) will be fast, and from user's perspective it is not query. But for any different content of `var` variable you should to get different queryid, that can be related to some slow queryid somewhere in Postgres. So if I have 200 names, then I get 200 different queryids, but it is not practical - so you use top one , or top ten

> 
> I didn't look thoroughly at plpgsql implementation yet, but I'm assuming that if we add an infrastructure to report some SPI call's queryid, it should only return the top level one, so the nested one will be automatically discarded by callers like plpgslq_check.

yes, but it can be saved to some plpgsql global variables - and plpgsql_check is called very immediately after statement fininshing, so global variables should be valid.

**Comment by rjuju** [comment]
I just pushed an additional commit that should fix all mentioned issues and also tries implement queryid computing for dynamic queries based on previous discussion as a basis for additional discussion.

**Comment by okbob** [comment]
Looks well, I have two comments

1. I dislike introduce new GUC. queryid should be displayed everytime. If queryid is not stable, then it should be filtered in regressions tests.

2. when you calculate queryid, probably there will be lot of allocations. I didn't check code, but "probably" better is using own memorycontext, then use it as default, and after all reset this memory context.

**Comment by rjuju** [comment]
Thanks for the feedback.  Clearly queryid won't be stable in regression tests as it depends on oids.  It also requires pg_stat_statements (or similar extension) to be activated, and I don't know if adding that depency for installcheck is wanted here.

1.  Unfortunately, contrib's "check" target, which spawns a new instance with an optional custom config file, isn't available for third-party extensions.  Should I add some new regression tests which check that queryid is not null?

2. indeed.  I'll add a new memory context for that!

**Comment by okbob** [comment]
> Thanks for the feedback. Clearly queryid won't be stable in regression tests as it depends on oids. It also requires pg_stat_statements (or similar extension) to be activated, and I don't know if adding that depency for installcheck is wanted here.
> 
>     1. Unfortunately, contrib's "check" target, which spawns a new instance with an optional custom config file, isn't available for third-party extensions.  Should I add some new regression tests which check that queryid is not null?
> 

REGRESS_OPTS = --temp-config $(top_srcdir)/contrib/pg_stat_statements/pg_stat_statements.conf

It's doesn't work?

Theoretically you can install own Post-parse-analysis hook and calculate own simple queryid - just some hash. This mode can be enabled by some GUC. Or ignore results, just do tests for ensuring so code should not crash, but results can be ignored.


>     2. indeed.  I'll add a new memory context for that!

**Comment by rjuju** [comment]
> > ```
> > 1. Unfortunately, contrib's "check" target, which spawns a new instance with an optional custom config file, isn't available for third-party extensions.  Should I add some new regression tests which check that queryid is not null?
> > ```
> 
> REGRESS_OPTS = --temp-config $(top_srcdir)/contrib/pg_stat_statements/pg_stat_statements.conf
> 
> It's doesn't work?

No, that's only suitable for `check` target which only works for in-tree extensions.  That's not usable with PGXS, see https://github.com/postgres/postgres/blob/master/src/makefiles/pgxs.mk#L430-L434.

`installcheck` use an existing server without any restart or configuration, so it's the user's duty to make sure that it's configured as required.

**Comment by rjuju** [comment]
> Theoretically you can install own Post-parse-analysis hook and calculate own simple queryid - just some hash. This mode can be enabled by some GUC. Or ignore results, just do tests for ensuring so code should not crash, but results can be ignored.

Yes, using a custom queryid could work, but it seems like a lot of trouble if that's only used for regression tests.  It could also be misleading for users, so I'd rather avoiding this solution.

We can either normalize the results, or provide an alternate expected output.  That would only ensure that we either have always a queryid or never, so I don't know if it's worth the trouble of maintaining 2 different expected files.

**Comment by okbob** [comment]
> > > ```
> > > 1. Unfortunately, contrib's "check" target, which spawns a new instance with an optional custom config file, isn't available for third-party extensions.  Should I add some new regression tests which check that queryid is not null?
> > > ```
> > 
> > 
> > REGRESS_OPTS = --temp-config $(top_srcdir)/contrib/pg_stat_statements/pg_stat_statements.conf
> > It's doesn't work?
> 
> No, that's only suitable for `check` target which only works for in-tree extensions. That's not usable with PGXS, see https://github.com/postgres/postgres/blob/master/src/makefiles/pgxs.mk#L430-L434.
> 
> `installcheck` use an existing server without any restart or configuration, so it's the user's duty to make sure that it's configured as required.

ok - because these numbers are calculated outside, I don't need to be tested here. Just do test so new patch is crash safe.

**Comment by okbob** [comment]
pá 13. 11. 2020 v 6:45 odesílatel Julien Rouhaud <notifications@github.com>
napsal:

> Theoretically you can install own Post-parse-analysis hook and calculate
> own simple queryid - just some hash. This mode can be enabled by some GUC.
> Or ignore results, just do tests for ensuring so code should not crash, but
> results can be ignored.
>
> Yes, using a custom queryid could work, but it seems like a lot of trouble
> if that's only used for regression tests. It could also be misleading for
> users, so I'd rather avoiding this solution.
>
> We can either normalize the results, or provide an alternate expected
> output. That would only ensure that we either have always a queryid or
> never, so I don't know if it's worth the trouble of maintaining 2 different
> expected files.
>
I think so custom query id with own post-parse-analyses hook is the best
solution - but the hook should be initialized by special plpgsql_check
function. I don't think so it can be problematic for users, because it will
be clean in tests too. Some like

select plpgsql_profiler_install_fake_queryid_hook();

profiler tests ...

select plpgsql_profiler_uninstall_fake_queryid_hook();

**Comment by rjuju** [comment]
I just pushed a new commit that should address your comments.  I also forgot to mention in the commit message, but while doing some more testing I found a bug that's now fixed with:
```
        SPIPlanPtr ptr  = expr->plan;
-       List *sources = SPI_plan_get_plan_sources(ptr);
+       List *sources;
+
+       if (!ptr)
+           return NOQUERYID;
+
+       sources = SPI_plan_get_plan_sources(ptr);
```

I'll continue to do some more testing locally.

**Comment by rjuju** [comment]
Great, thanks a lot for the extensive help!  I just pushed a new commit to add some documentation on the new feature.  While at at I added some other fixes I could spot.

**Comment by okbob** [comment]
Merged, thank you very much for patch. I'll release new version early. Please, can you do some blog post about this feature?

**Comment by rjuju** [comment]
Thanks a lot!  Sure, I'll write a blog post about this new feature.

Before doing a new release, there's another change I'd be interested in, which is making MAX_SHARED_CHUNKS configurable via a new PGC_POSTMASTER GUC.  I'll open a new issue to discuss the details.

**Comment by okbob** [comment]
ok
